### PR TITLE
Loading nn.Module from checkpoint tutorial (non ghstack)

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Added tutorial for some utilities for loading checkpoint 

1.  The ``mmap`` keyword argument on ``torch.load``
2.  The ``torch.device()`` context manager
3.  The ``assign`` keyword argument on ``nn.Module.load_state_dict()``

### CODE REVIEW

File: recipes_source/recipes/module_load_state_dict_tips.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
+"""
+
+Tips for Loading an ``nn.Module`` from a Checkpoint
+===================================================
+
+If you're loading a checkpoint and want to reduce compute and memory as much as possible,
+this tutorial shares some recommended practices. In particular, we will discuss
```
Comment (svekars) [inline recipes_source/recipes/module_load_state_dict_tips.py:7]:
```suggestion
this tutorial shares some recommended practices. In particular, we will discuss the following:
```

File: recipes_source/recipes/module_load_state_dict_tips.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+2.  The ``torch.device()`` context manager
+3.  The ``assign`` keyword argument on ``nn.Module.load_state_dict()``
+
+.. note::
+   This recipe requires PyTorch 2.1.0 or later.
+"""
+
+
+########################################
+# Let us consider a simple ``nn.Module``
```
Comment (svekars) [inline recipes_source/recipes/module_load_state_dict_tips.py:19]: Would be good add a bit more description here about what the code below does.

File: recipes_source/recipes/module_load_state_dict_tips.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+    def forward(self, x):
+        return self.linears(x)
+
+
+m = SomeModule(1000)
+torch.save(m.state_dict(), 'checkpoint.pth')
+
+#################################################################
+# The follow snippet demonstrates the use of the three utilities.
```
Comment (svekars) [inline recipes_source/recipes/module_load_state_dict_tips.py:37]: can you list the utilities.

File: recipes_source/recipes/module_load_state_dict_tips.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+state_dict = torch.load('checkpoint.pth', mmap=True)
+with torch.device('meta'):
+  meta_m = SomeModule(1000)
+meta_m.load_state_dict(state_dict, assign=True)
+
+#############################################################################
+# Taking a step back, let us inspect the following more vanilla code snippet
+# that does not use any of the features listed above:
```
Comment (svekars) [inline recipes_source/recipes/module_load_state_dict_tips.py:46]:
```suggestion
# Compare the snippet below to the one above:
```

File: recipes_source/recipes/module_load_state_dict_tips.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+state_dict = torch.load('checkpoint.pth')
+m = SomeModule(1000)
+m.load_state_dict(state_dict)
+
```
Comment (svekars) [inline recipes_source/recipes/module_load_state_dict_tips.py:51]:
```suggestion
# The second example does not use any of the features listed above.
```
Can you add a conclusion about this comparison?

File: recipes_source/recipes/module_load_state_dict_tips.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+# First let us consider what happens when we ``torch.load`` the checkpoint.
+# At ``torch.save`` time, tensor storages are tagged with the device they are
+# saved on. At ``torch.load`` time, tensor storages will be loaded to the device
+# they were tagged with (unless this behavior is overridden using the
+# ``map_location`` flag). For ease of explanation, let us assume that the tensors
+# were saved on CPU. This means that on the first line all tensor storages will be
+# loaded into CPU RAM, which can be undesirable when
+#     1. CPU RAM is smaller than the size of the checkpoint
+#     2. Waiting for the entire checkpoint to be loaded into RAM before
+#        doing for example some per-tensor processing
```
Comment (svekars) [inline recipes_source/recipes/module_load_state_dict_tips.py:64]:
```suggestion
# First, let us consider what happens when we load the checkpoint with``torch.load``.
# When we save a checkpoint with ``torch.save``, tensor storages are tagged with the device they are
# saved on. With ``torch.load``, tensor storages will be loaded to the device
# they were tagged with (unless this behavior is overridden using the
# ``map_location`` flag). For ease of explanation, let us assume that the tensors
# were saved on CPU. This means that on the first line all tensor storages will be
# loaded into CPU RAM, which can be undesirable when:
#
# * CPU RAM is smaller than the size of the checkpoint.
# * Waiting for the entire checkpoint to be loaded into RAM before
#    performing, for example, some per-tensor processing.
```

File: recipes_source/recipes/module_load_state_dict_tips.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+# storages will be memory-mapped.
+
+start_time = time.time()
+state_dict = torch.load('checkpoint.pth', mmap=True)
+end_time = time.time()
+print(f"loading time with mmap={end_time - start_time}")
+
+######################################################################################
+# As mentioned above, one can use this argument to do per-tensor processing on a
+# checkpoint without loading all tensor storages into CPU memory upfront. For example,
```
Comment (svekars) [inline recipes_source/recipes/module_load_state_dict_tips.py:86]:
```suggestion
# As mentioned above, one can use this argument to do per-tensor processing on a
# checkpoint without loading all tensor storages into CPU memory upfront. For example:
```

File: recipes_source/recipes/module_load_state_dict_tips.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    return processed_t
+
+for key in state_dict.keys():
+    device = torch.device('cuda:' + str(int(key.lstrip("linears.")[0]) % 8))
+    state_dict[key] = my_processing_function(key, device)
+
+##############################################
+# Using ``torch.device('meta')``
+# ------------------------------
+# Next, we consider the creation of the module.
```
Comment (svekars) [inline recipes_source/recipes/module_load_state_dict_tips.py:104]:
```suggestion
# Next, let's consider the creation of the module.
```

File: recipes_source/recipes/module_load_state_dict_tips.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+#        the saved state dictionary also allocates memory for the parameters/buffers in the checkpoint.
+#
+# In order to solve these two problems, we can use the ``torch.device()``
+# context manager with ``device='meta'`` when we instantiate the ``nn.Module()``.
+#
+# The `torch.device() <https://pytorch.org/docs/main/tensor_attributes.html#torch-device>`_
+# context manager makes sure that factory calls will be performed as if they
+# were passed the specified ``device`` as an argument. Tensors on ``torch.device('meta')`` do not
+# carry data. However, they possess all other metadata a tensor carries such as ``.size()``, ``.stride()``
+# and ``.requires_grad`` etc.
```
Comment (svekars) [inline recipes_source/recipes/module_load_state_dict_tips.py:123]:
```suggestion
# is wasteful when we want to load a checkpoint as one of the following:
# * The result of the initialization kernels will be overwritten by ``load_state_dict()``
#    without ever being used, so initialization is wasteful.
# * We are allocating memory for these parameters/buffers in RAM while ``torch.load`` of
#    the saved state dictionary also allocates memory for the parameters/buffers in the checkpoint.
#
# In order to solve these two problems, we can use the ``torch.device()``
# context manager with ``device='meta'`` when we instantiate the ``nn.Module()``.
#
# The `torch.device() <https://pytorch.org/docs/main/tensor_attributes.html#torch-device>`_
# context manager makes sure that factory calls will be performed as if they
# were passed the specified ``device`` as an argument. Tensors on ``torch.device('meta')`` do not
# carry data. However, they possess all other metadata a tensor carries such as ``.size()``, ``.stride()``,
# ``.requires_grad``, and other.
```

File: recipes_source/recipes/module_load_state_dict_tips.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+# --------------------------------------
+# Next, we consider the loading of the state dictionary.
+
+m.load_state_dict(state_dict)
+
+###############################################################################
+# ``nn.Module.load_state_dict()`` is usually implemented via an in-place
+# ``param_in_model.copy_(param_in_state_dict)`` (i.e. a copy from the
+# parameter/buffer with the corresponding key in the state dictionary into
+# the parameter/buffer in the ``nn.Module``).
```
Comment (svekars) [inline recipes_source/recipes/module_load_state_dict_tips.py:138]:
```suggestion
# ``param_in_model.copy_(param_in_state_dict)``. For example, a copy from the
# parameter/buffer with the corresponding key in the state dictionary into
# the parameter/buffer in the ``nn.Module``.
```

File: recipes_source/recipes/module_load_state_dict_tips.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+# A caveat here is that since optimizers hold a reference to
+# ``nn.Module.parameters()``, the optimizer must be initialized after the module
+# is loaded from state dict if ``assign=True`` is passed.
+
+new_m.load_state_dict(state_dict, assign=True)
+# This MUST be done AFTER the load_state_dict with assign.
+opt = torch.optim.SGD(new_m.parameters(), lr=1e-3)
+
+###############################################################################
+# To recap, in this tutorial we learned about ``torch.load(mmap=True)``, the
```
Comment (svekars) [inline recipes_source/recipes/module_load_state_dict_tips.py:153]:
```suggestion
# Conclusion
# ------------
# 
# To recap, in this tutorial we learned about ``torch.load(mmap=True)``, the
```

File: recipes_source/recipes/module_load_state_dict_tips.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+# ``torch.device()`` context manager with ``device=meta`` and
```
Comment (svekars) [inline recipes_source/recipes/module_load_state_dict_tips.py:154]:
```suggestion
# ``torch.device()`` context manager with ``device=meta``, and
```

**Review by svekars — COMMENTED** [review]
thanks, @mikaylagawarecki. Looks good, just a couple of editorial suggestions.

**Review by albanD — APPROVED** [review]
SGTM

**Review by svekars — APPROVED** [review]

**Pull request closed without merging**